### PR TITLE
feat(propagation): add baggage support to http propagation

### PR DESCRIPTION
**Pull request by HovhannesV**
### Goal
Goal is to have baggage item support with http propagation, which will be compatible with other language DataDog tracing libs e.g. java, nodejs

### Changes
- Reused codebase from this [PR](https://github.com/DataDog/dd-trace-py/pull/3020) which had conflicts with main branch.
- Added http propagator extract functionality and edited corresponding test

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](../docs/contributing.rst#release-branch-maintenance))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](../docs/contributing.rst#release-branch-maintenance)

### CODE REVIEW

**Review by majorgreys — CHANGES_REQUESTED** [review]
@HovhannesV Thank you for initiating this effort and I apologize it took us this long to reply. If you are still interested in carrying this PR forward, we will assist in getting it merged. I went ahead and resolved merge conflicts.

My only requested change would be to have this behind a feature flag so a library user would have to opt-in. And, as such, we should avoid changing the public API. What are your thoughts?

We have been focused on adding support for the OpenTelemetry API to the library, though for now its Baggage API is not supported. When we do begin that work, I'll be sure to ping you for your feedback.

**Comment by HovhannesV** [comment]
@majorgreys Thanks for your insight. Yes, I would like to have this merged if thats ok from your side.

I am not sure I understand whats the feature flag going to control. Did you mean to control automatic http propagation ? 
If yes, at the time of development I have checked the Node JS and Java versions of the lib and it was automatically propagating the baggage headers. 
Also not sure I got the Public API part, as the changes are backward compatible. Could you please elaborate more.

P.S.

Yes, I would be happy to take a look at the upcoming OpenTelemetry changes.

**Comment by majorgreys** [comment]
@HovhannesV, thank you for the quick reply and offering to carry the PR to completion.

> I am not sure I understand whats the feature flag going to control. Did you mean to control automatic http propagation ? 

Right, the changes to the extract/inject would be behind a new option added to the global configuration object `ddtrace.config`, with an environment variable like `DD_TRACE_PROPAGATION_HTTP_BAGGAGE_ENABLED`. 

> If yes, at the time of development I have checked the Node JS and Java versions of the lib and it was automatically propagating the baggage headers. 

We in the Python library try to introduce new features whenever possible behind a feature flag. This avoids customers being opted into new features, which has them and us many headaches in the past. In this instance, uses who want baggage support can enable it and those that do not can avoid it entirely until the feature has been tested and stable.

Do you think that's a problem for this specific change?

> Also not sure I got the Public API part, as the changes are backward compatible. Could you please elaborate more.

Sorry, I should have clarified earlier. The `Span` class is in our public API. I will suggest the change in the pull request to make it clear.

File: ddtrace/span.py [hunk 1]
```diff
@@ -446,6 +446,19 @@ def get_metric(self, key: _TagNameType) -> Optional[NumericType]:
         """Return the given metric or None if it doesn't exist."""
         return self._metrics.get(key)
 
+    def set_baggage_item(self, key, value):
```
Comment (majorgreys) [inline ddtrace/span.py:449]:
```suggestion
    def _set_baggage_item(self, key, value):
```

File: ddtrace/span.py [hunk 2]
```diff
@@ -446,6 +446,19 @@ def get_metric(self, key: _TagNameType) -> Optional[NumericType]:
...
+        # type: (str, Any) -> Span
+        """Sets a baggage item in the span context of this span.
+        Baggage is used to propagate state between spans (in-process, http/https).
+        """
+        self._context = self.context.with_baggage_item(key, value)
+        return self
+
+    def get_baggage_item(self, key):
```
Comment (majorgreys) [inline ddtrace/span.py:457]:
```suggestion
    def _get_baggage_item(self, key):
```

File: tests/tracer/test_span.py [hunk 3]
```diff
@@ -96,6 +96,23 @@ def test_set_tag_bool(self):
...
+        span1.set_baggage_item("item1", "123")
+
+        span2 = Span(name="test.span2", context=span1.context)
+        span2.set_baggage_item("item2", "456")
+
+        assert span2.get_baggage_item("item1") == "123"
+        assert span2.get_baggage_item("item2") == "456"
+        assert span1.get_baggage_item("item1") == "123"
+        assert span1.get_baggage_item("item2") is None
+
```
Comment (majorgreys) [inline tests/tracer/test_span.py:115]:
```suggestion
    def test_set_baggage_item(self):
        s = Span(name="test.span")
        s._set_baggage_item("custom.key", "123")
        assert s._get_baggage_item("custom.key") == "123"

    def test_baggage_propagation(self):
        span1 = Span(name="test.span1")
        span1._set_baggage_item("item1", "123")

        span2 = Span(name="test.span2", context=span1.context)
        span2._set_baggage_item("item2", "456")

        assert span2._get_baggage_item("item1") == "123"
        assert span2._get_baggage_item("item2") == "456"
        assert span1._get_baggage_item("item1") == "123"
        assert span1._get_baggage_item("item2") is None

```

**Review by majorgreys — COMMENTED** [review]
We should avoid a breaking change to the public API.

**Review by majorgreys — CHANGES_REQUESTED** [review]
Great! Similarly, we want the `Context` class not to have any public api changes.

**Comment by majorgreys** [comment]
> I think I cannot make the `get_baggage_item` method private on Context class, as when the flag `propagation_http_baggage_enabled` is turned on users should be able to read the baggage items with code.

@HovhannesV The method would be private by convention only since Python would not disallow accessing the method on the `Context` instance:

```
>>> class A:
...     def _private(self):
...             pass
...
>>> a = A()
>>> a._private()
```

Does that help clarify my suggestions?

**Comment by HovhannesV** [comment]
@majorgreys Thanks for your clarification. Actually my question was also about convention, too. I was thinking that users will be confused when they won't be able to find some Public API for working with baggage items. But if you think thats ok, I will change it then.

**Comment by HovhannesV** [comment]
@majorgreys Seems like I have done the changes you requested, kindly asking you to review again.

**Review by majorgreys — CHANGES_REQUESTED** [review]
@HovhannesV thank you! Looks super close, just one remaining request with regards to the public interface changes.

Also, we will need a release note for this. Can you add one with how to use this feature and remark that it is currently private and subject to change? 

https://ddtrace.readthedocs.io/en/stable/releasenotes.html

**Comment by HovhannesV** [comment]
@majorgreys  Thanks for the follow up. Hope I did all the changes you requested. Kindly asking you to review again.

File: releasenotes/notes/baggage-support-with-http-propagation-e00d106eefff67f1.yaml [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    Datadog Tracing adds support for OpenTracing Baggage Items with HTTP Propagation.
+  
+    .. note::
+    Methods for working with Baggage Items programmatically are private and subject to change.
+features:
+  - |
+    tracing: This introduces support for OpenTracing Baggage Items with HTTP Propagation.
+    Enable HTTP Propagation by setting ``DD_TRACE_PROPAGATION_HTTP_BAGGAGE_ENABLED`` to true.
+    Get/Set baggage item by referring to ``_set_baggage_item`` and ``_get_baggage_item`` methods on ``Context`` class
+
```
Comment (majorgreys) [inline releasenotes/notes/baggage-support-with-http-propagation-e00d106eefff67f1.yaml:12]:
```suggestion
features:
  - |
    tracing: Introduces support for OpenTracing Baggage Items with HTTP Propagation. Enable this support by ``DD_TRACE_PROPAGATION_HTTP_BAGGAGE_ENABLED=true``. The ``Context._set_baggage_item`` and ``Context._get_baggage_item`` internal methods are provided for manual modifications to the Baggage Items. These API changes are subject to change.

```

**Review by majorgreys — CHANGES_REQUESTED** [review]
Small changes to the release notes, otherwise LGTM!

**Comment by HovhannesV** [comment]
I hope we can merge it now ? :)

**Review by majorgreys — APPROVED** [review]

**Pull request merged**